src/firmware/main/app/control.c

```diff
@@ -0,0 +1,184 @@
+ *         accelerate the robot as fast as physically possible
+ */
+float app_control_getMaximalAccelScaling(const FirmwareRobot_t* robot,
+                                         const float linear_accel_x,
+                                         const float linear_accel_y, float angular_accel)
+{
+    const RobotConstants_t robot_constants = app_firmware_robot_getRobotConstants(robot);
+
+    // first convert accelerations into consistent units
+    // choose units of Force (N)
+    float normed_force[3];
+    normed_force[0] = linear_accel_x * robot_constants.mass;
+    normed_force[1] = linear_accel_y * robot_constants.mass;
+    normed_force[2] =
+        angular_accel * robot_constants.moment_of_inertia / robot_constants.robot_radius;
+
+    float wheel_forces[4];
+    force3_to_force4(normed_force, wheel_forces);
+
+    Wheel_t* wheels[4];
+    wheels[0] = app_firmware_robot_getFrontLeftWheel(robot);
+    wheels[1] = app_firmware_robot_getBackLeftWheel(robot);
+    wheels[2] = app_firmware_robot_getBackRightWheel(robot);
+    wheels[3] = app_firmware_robot_getFrontRightWheel(robot);
+
+    float battery_voltage = app_firmware_robot_getBatteryVoltage(robot);
+
+    return app_control_getMaximalTorqueScaling(wheels, wheel_forces, battery_voltage);
+}
+
+void app_control_applyAccel(FirmwareRobot_t* robot, float linear_accel_x,
+                            float linear_accel_y, float angular_accel)
+{
+    const RobotConstants_t robot_constants = app_firmware_robot_getRobotConstants(robot);
+
+    // check for max acceleration in direction of the vel difference
+    float scaling = app_control_getMaximalAccelScaling(robot, linear_accel_x,
+                                                       linear_accel_y, angular_accel);
+
+    // if the naive 1 tick acceleration violates the limits of the robot
+    // scale it to maximum
+    // if the 1 tick acceleration is below the limit, then leave it
+    if (scaling < 1.0f)
+    {
+        linear_accel_x *= scaling;
+        linear_accel_y *= scaling;
+        angular_accel *= scaling;
+    }
+
+    ControllerState_t* controller_state = app_firmware_robot_getControllerState(robot);
+    float prev_linear_accel_x           = controller_state->last_applied_acceleration_x;
+    float prev_linear_accel_y           = controller_state->last_applied_acceleration_y;
+    float prev_angular_accel = controller_state->last_applied_acceleration_angular;
+
+    float linear_diff_x = linear_accel_x - prev_linear_accel_x;
+    float linear_diff_y = linear_accel_y - prev_linear_accel_y;
+    float angular_diff  = angular_accel - prev_angular_accel;
+
+    const float jerk_limit                       = robot_constants.jerk_limit;
+    const float linear_acceleration_change_limit = robot_constants.jerk_limit * TICK_TIME;
+    const float angular_acceleration_change_limit =
+        jerk_limit / ROBOT_RADIUS * TICK_TIME * 5.0f;
```

Comment:
not sure if in scope but what is this magic number 5.0f?

Reply:
Not sure, leaving as out-of-scope for now. There will definitely need to be some more cleanup after this........